pkg/v1/remote/transport/ping.go

```diff
@@ -71,12 +71,18 @@ func parseChallenge(suffix string) map[string]string {
 func ping(ctx context.Context, reg name.Registry, t http.RoundTripper) (*pingResp, error) {
 	client := http.Client{Transport: t}
 
-	// This first attempts to use "https" for every request, falling back to http
-	// if the registry matches our localhost heuristic or if it is intentionally
-	// set to insecure via name.NewInsecureRegistry.
-	schemes := []string{"https"}
-	if reg.Scheme() == "http" {
-		schemes = append(schemes, "http")
+	// If it is intentionally set to insecure via name.NewInsecureRegistryOnly, then only
+	// try to use "http" for the request.
+	// Otherwise, this first attempts to use "https" for every request, falling back to http
+	// if the registry matches our localhost heuristic.
+	var schemes []string
+	if reg.IsInsecure() {
+		schemes = []string{"http"}
```

Comment:
I list all conditions below to figure out how this patch will make a difference to the usage.

|No.|Registry Protocols|Insecure Flag|Match localhost pattern|Protocol Used Before|Protocol Used After|
|  ----  | ----  | ----  | ----  | ---- | ---- |
|1|HTTPS+HTTP|✖|✖|HTTPS|HTTPS|
|2|HTTPS+HTTP|✖|✔|HTTPS|HTTPS|
|3|HTTPS+HTTP|✔|✖|HTTPS|HTTP|
|4|HTTPS+HTTP|✔|✔|HTTPS|HTTP|
|5|HTTPS-only|✖|✖|HTTPS|HTTPS|
|6|HTTPS-only|✖|✔|HTTPS|HTTPS|
|7|HTTPS-only|✔|✖|HTTPS|failed|
|8|HTTPS-only|✔|✔|HTTPS|failed|
|9|HTTP-only|✖|✖|failed|failed|
|10|HTTP-only|✖|✔|HTTP|HTTP|
|11|HTTP-only|✔|✖|HTTP|HTTP|
|12|HTTP-only|✔|✔|HTTP|HTTP|

The conditions on which performances will be different are No.7/No.8. On these conditions, users just need to unset the insecure flag to make it work. Considering that the insecure flag is default `False`, these conditions may not happen.

Reply:
> Considering that the insecure flag is default False, these conditions may not happen.

My concern is that this option might be set "globally" for a given consumer. If you look at kaniko, they have [a bunch of flags](https://github.com/GoogleContainerTools/kaniko#--insecure) that are booleans to enable insecure behavior.

You can imagine someone with a Dockerfile that pulls FROM two different registries: their secure production registry, and an insecure registry running on their network.

If we don't have a fallback to https, pulling from their secure registry would just fail.

Arguably, any decent web server will redirect from http to https, so this might not be a problem in practice, but I wanted to bring it up as a potentially breaking change.

Reply:
I agreed that maybe some kaniko users enbale the `insecure` flag and have secure/insecure registries in one Dockerfile. On that condition, the pulling from thesecure registries would fail.  However, in kaniko, users can also use the `insecure-registry` to config which registries are insecure and the others will still use HTTPS for pulling and pushing.

In my opinion, it is determined that if users set `insecure`, then we will try `HTTP` at first. And whether or not we fallback to HTTPS is another consideration.

Reply:
I think `fallback to HTTPS after HTTP is failed` is reasonable for this issue.